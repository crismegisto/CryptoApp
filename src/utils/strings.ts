export const isStringMatch = (query: string | undefined, target: string): boolean => {
  const queryLowerCase = query?.toLowerCase();
  return queryLowerCase ? target.toLowerCase().includes(queryLowerCase) : false;
};
