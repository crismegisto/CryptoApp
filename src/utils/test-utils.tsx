import React, { PropsWithChildren } from 'react';
import { render, RenderOptions } from '@testing-library/react-native';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from '../store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState), ...renderOptions }: ExtendedRenderOptions = {}
) {
  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => (
    <Provider store={store}>{children}</Provider>
  );

  return { store, ...render(ui, { wrapper, ...renderOptions }) };
}

export * from '@testing-library/react-native';

export { renderWithProviders as render };
