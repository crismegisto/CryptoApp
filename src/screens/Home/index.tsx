import { SafeAreaView } from 'react-native';
import React, { useMemo, useState } from 'react';
import { FlashList, ListRenderItem } from '@shopify/flash-list';

import { useFetchCoinsQuery } from '../../services/apiSlice';
import CoinCard from '../../components/CoinCard';
import { Coin, ListKeyExtractor } from '../../interfaces';
import SearchBar from '../../components/SearchBar';
import { isStringMatch } from '../../utils/strings';
import styles from './styles';

const Home = () => {
  const { data: coins = [] } = useFetchCoinsQuery();
  const [query, setQuery] = useState('');

  const filteredCoins = useMemo<Coin[]>(
    () => coins.filter(({ symbol, name }) => isStringMatch(query, symbol) || isStringMatch(query, name)),
    [coins, query]
  );

  const renderItem: ListRenderItem<Coin> = ({ item }) => <CoinCard {...item} />;

  const keyExtractor: ListKeyExtractor<Coin> = ({ id }) => String(id);
  return (
    <SafeAreaView style={styles.container}>
      <SearchBar onSearch={(text: string | null) => setQuery(String(text))} />
      <FlashList
        renderItem={renderItem}
        keyExtractor={keyExtractor}
        data={query ? filteredCoins : coins}
        estimatedItemSize={100}
      />
    </SafeAreaView>
  );
};

export default Home;
