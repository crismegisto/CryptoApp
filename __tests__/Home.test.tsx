import React from 'react';

import Home from '../src/screens/Home';
import { render, fireEvent, waitFor, act, screen } from '../src/utils/test-utils';

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn()
    })
  };
});

jest.useFakeTimers();
describe('Home testing', () => {
  test('snapshot', async () => {
    render(<Home />);
    expect(screen.toJSON()).toMatchSnapshot();
    expect(await screen.findByText('Bitcoin')).toBeTruthy();
  });

  test('renders coin cards correctly', async () => {
    const { findAllByTestId } = render(<Home />);
    const renderedCoinCards = await findAllByTestId('coin-card');
    expect(renderedCoinCards).toHaveLength(2);
  });

  test('filters coins based on search input', async () => {
    const { getByTestId, findAllByTestId } = render(<Home />);
    const searchBar = getByTestId('input');

    await act(async () => {
      fireEvent.changeText(searchBar, 'SHIB');
      await waitFor(() => expect(searchBar.props.value).toBe('SHIB'));

      jest.runAllTimers();
      const filteredCoins = await findAllByTestId('coin-card');
      expect(filteredCoins).toHaveLength(1);
    });
  });
});
