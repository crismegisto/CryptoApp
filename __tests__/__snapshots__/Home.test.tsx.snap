// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home testing snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "white",
        "borderColor": "green",
        "borderRadius": 10,
        "borderWidth": 1,
        "margin": 12,
        "padding": 10,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Search..."
      testID="input"
      value=""
    />
  </View>
  <View
    style={
      {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RCTScrollView
      applyWindowCorrection={[Function]}
      canChangeSize={true}
      contentContainerStyle={
        {
          "backgroundColor": undefined,
          "minHeight": 1,
          "minWidth": 1,
          "paddingLeft": 0,
          "paddingRight": 0,
        }
      }
      contentHeight={0}
      contentWidth={400}
      data={[]}
      dataProvider={
        DataProvider {
          "_data": [],
          "_firstIndexToProcess": 0,
          "_hasStableIds": true,
          "_requiresDataChangeHandling": false,
          "_size": 0,
          "getStableId": [Function],
          "rowHasChanged": [Function],
        }
      }
      disableRecycling={false}
      estimatedItemSize={100}
      extendedState={{}}
      externalScrollView={[Function]}
      finalRenderAheadOffset={250}
      forceNonDeterministicRendering={true}
      horizontal={false}
      initialOffset={0}
      initialRenderIndex={0}
      isHorizontal={false}
      keyExtractor={[Function]}
      layoutProvider={
        GridLayoutProviderWithProps {
          "_acceptableRelayoutDelta": 1,
          "_getHeightOrWidth": [Function],
          "_getLayoutTypeForIndex": [Function],
          "_getSpan": [Function],
          "_hasExpired": false,
          "_isHorizontal": false,
          "_lastLayoutManager": GridLayoutManager {
            "_acceptableRelayoutDelta": 1,
            "_getSpan": [Function],
            "_isGridHorizontal": false,
            "_isHorizontal": false,
            "_layoutProvider": [Circular],
            "_layouts": [],
            "_maxSpan": 1,
            "_renderWindowSize": {
              "height": 900,
              "width": 400,
            },
            "_totalHeight": 0,
            "_totalWidth": 400,
            "_window": {
              "height": 900,
              "width": 400,
            },
          },
          "_maxSpan": 1,
          "_renderWindowSize": {
            "height": 900,
            "width": 400,
          },
          "_setLayoutForType": [Function],
          "_tempDim": {
            "height": 0,
            "width": 0,
          },
          "averageWindow": AverageWindow {
            "currentAverage": 100,
            "currentCount": 1,
            "inputValues": [
              100,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
              ,
            ],
            "nextIndex": 1,
          },
          "defaultEstimatedItemSize": 100,
          "layoutObject": {
            "size": undefined,
            "span": undefined,
          },
          "props": {
            "data": [],
            "estimatedItemSize": 100,
            "keyExtractor": [Function],
            "numColumns": 1,
            "renderItem": [Function],
          },
          "renderWindowInsets": {
            "height": 0,
            "width": 0,
          },
          "shouldRefreshWithAnchoring": true,
        }
      }
      maxRenderAhead={750}
      numColumns={1}
      onEndReached={[Function]}
      onEndReachedThreshold={0}
      onEndReachedThresholdRelative={0}
      onItemLayout={[Function]}
      onLayout={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onSizeChanged={[Function]}
      onVisibleIndicesChanged={[Function]}
      removeClippedSubviews={false}
      renderAheadOffset={0}
      renderAheadStep={250}
      renderContentContainer={[Function]}
      renderItem={[Function]}
      renderItemContainer={[Function]}
      rowRenderer={[Function]}
      scrollEventThrottle={16}
      scrollThrottle={16}
      scrollViewProps={
        {
          "contentContainerStyle": {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 0,
            "paddingRight": 0,
          },
          "onLayout": [Function],
          "onScrollBeginDrag": [Function],
          "refreshControl": undefined,
          "style": {
            "minHeight": 1,
            "minWidth": 1,
          },
        }
      }
      style={
        {
          "minHeight": 1,
          "minWidth": 1,
        }
      }
      suppressBoundedSizeException={true}
      windowCorrectionConfig={
        {
          "applyToInitialOffset": true,
          "applyToItemScroll": true,
          "value": {
            "endCorrection": 0,
            "startCorrection": 0,
            "windowShift": -0,
          },
        }
      }
    >
      <View>
        <View
          style={
            {
              "flexDirection": "column",
            }
          }
        >
          <View
            style={
              {
                "paddingLeft": 0,
                "paddingTop": 0,
              }
            }
          />
          <View
            style={
              [
                undefined,
                undefined,
              ]
            }
          />
          <AutoLayoutView
            enableInstrumentation={false}
            horizontal={false}
            onBlankAreaEvent={[Function]}
            onLayout={[Function]}
            renderAheadOffset={0}
            scrollOffset={0}
            style={
              {
                "height": 0,
                "width": 400,
              }
            }
            windowSize={900}
          />
          <CellContainer
            index={-1}
            style={
              [
                undefined,
                undefined,
              ]
            }
          />
          <View
            style={
              {
                "paddingBottom": 0,
                "paddingRight": 0,
              }
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
